bundles/org.openhab.binding.myq/src/main/java/org/openhab/binding/myq/internal/handler/MyQAccountHandler.java

```diff
@@ -0,0 +1,329 @@
+        normalPollFuture = null;
+    }
+
+    private synchronized void stopRapidPoll() {
+        stopFuture(rapidPollFuture);
+        rapidPollFuture = null;
+    }
+
+    private void stopFuture(@Nullable Future<?> future) {
+        if (future != null) {
+            future.cancel(true);
+        }
+    }
+
+    private synchronized void restartPolls(boolean rapid) {
+        stopPolls();
+        if (rapid) {
+            normalPollFuture = scheduler.scheduleWithFixedDelay(this::normalPoll, 35, normalRefreshSeconds,
+                    TimeUnit.SECONDS);
+            rapidPollFuture = scheduler.scheduleWithFixedDelay(this::rapidPoll, 3, RAPID_REFRESH_SECONDS,
+                    TimeUnit.SECONDS);
+        } else {
+            normalPollFuture = scheduler.scheduleWithFixedDelay(this::normalPoll, 0, normalRefreshSeconds,
+                    TimeUnit.SECONDS);
+        }
+    }
+
+    private void normalPoll() {
+        stopRapidPoll();
+        fetchData();
+    }
+
+    private void rapidPoll() {
+        fetchData();
+    }
+
+    private synchronized void fetchData() {
+        try {
+            if (securityToken == null) {
+                login();
+                if (securityToken != null) {
+                    getAccount();
+                }
+            }
+
+            if (securityToken != null) {
+                getDevices();
+            }
+        } catch (InterruptedException e) {
+        }
+    }
+
+    private void login() throws InterruptedException {
+        HttpResult result = sendRequest(BASE_URL + "/v5/Login", HttpMethod.POST, null,
+                new StringContentProvider(gsonUpperCase.toJson(new LoginRequestDTO(username, password))),
+                "application/json");
+        if (HttpStatus.isSuccess(result.responseCode)) {
+            LoginResponseDTO loginResponse = gsonUpperCase.fromJson(result.content, LoginResponseDTO.class);
+            if (loginResponse != null) {
+                securityToken = loginResponse.securityToken;
+                updateStatus(ThingStatus.ONLINE);
+            } else {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Invalid Response Body");
+            }
+        } else {
+            handleErrorResponse(result);
+            if (thing.getStatusInfo().getStatusDetail() == ThingStatusDetail.CONFIGURATION_ERROR) {
+                // bad credentials, stop trying to login
+                stopPolls();
+            }
+        }
+    }
+
+    private void getAccount() throws InterruptedException {
+        HttpResult result = sendRequest(BASE_URL + "/v5/My?expand=account", HttpMethod.GET, securityToken, null, null);
+        if (HttpStatus.isSuccess(result.responseCode)) {
+            account = gsonUpperCase.fromJson(result.content, AccountDTO.class);
+        } else {
+            handleErrorResponse(result);
+        }
+    }
+
+    private void getDevices() throws InterruptedException {
```

Comment:
If I see correctly this method does the (background) discovery of devices. So, it's better to move this into the discovery service. Then, you won't have a problem with the depdendencies of handler and discovery service either.

Reply:
So, when i poll the API i get all devices, so every poll cycle (60 seconds default ) i have the opportunity to discover new devices. I assumed this would be more efficient than having to schedule a separate call to the bridge to do an additional poll just to discover devices, that seemed less efficient with more complexity.

Reply:
Maybe the API has the ability to query only the state of one devices at a time? If not, I would do a discovery only if the user triggered a scan. That would reduce overall code complexity.

Reply:
Its' a single api call that i'm using, the API is not documented and heavily rate limited, its not clear even if there is a single device endpoint (which i have not seen), if you had more then one, how that would blow through the rate limiting polling for each.  FYI i have a 2 bay garage, so i have 2 units under the same account.  

>That would reduce overall code complexity.

I'm all for reducing complexity, but i'm not sure i see that happening, we would have the same discovery code that exists, PLUS i would need to add additional methods to call the same poll code, but this time not to update devices, just for discovery,  its going to be more code, additional coordination of outgoing requests and much more complex IMHO.

Reply:
It's reasonable to make one request for all devices due to the rate limit. Nevertheless, the discovery service could make its own request as it is not triggered permanently. This way you could cut off the counter intuitive reverse dependency of bridge handler and discovery service and you can implement the ThingHandlerService. Of course you could make `getDevices()` stateless and reuseable to use it in the bridge handler and the discovery service.

Reply:
Hi @fwolter  i have refactored the discover class to use the ThingHandlerService .  I did not realize this removed the need for all the boiler plate code in the handler factory to create, track and destroy the discovery service, so i agree this was worth the refactoring.

Reply:
Great!